### PR TITLE
Upstream nRF54L15 building support and sync the repos

### PR DESCRIPTION
**Pull request by Vge0rge**
This is an effort to sync our fork more with upstream TF-M and to add support for building upstream TF-M with nRF54L15.

In order to support the nRF54L15 I copied some configuration files from nRF5340 and modified them to be able to build. These files are not used in sdk-nrf, they are only used in upstream TF-M. 

Here I am also trying to sync the repos in general which is in essence a mini upmerge.

My methodology was:
I reverted all the commits which start with an [nrf...] tag.
I applied all the [nrf fromtree] commits that we had.
I reverted the Zephyrs fork only commits which affected my changes.
Then I manually  cherry picked all the noups that need to go upstream and then I finally squashed them in one commit. 
Finally I applied all the noup commits which need to stay noup. 

When I cherry picked the noup commits I had to resolve conflicts either in place or in separate commits.  The main reasoning for needing to have separate commits to resolve conflicts is that the [nrf fromlist] commit needs to be applied cleanly upstream and here in our fork. 

I manually verified the diff of the folders between the current main branch here and  this PR and I see only minimal changes which should not affect anything.

The fromlist commit in this PR contains the following commits, and additionally it contains changed relevant to conflict resolution.
[nrf noup] platform: Disable default RTC30 IRQ handler
[nrf noup] platform: nordic_nrf: make RRAM writes unbuffered
[nrf noup] platform: nordic_nrf: 54l15: clean up RRAM write buffer size setting
[nrf noup] platform: nordic_nrf: 54l15: do not "erase" the flash
[nrf noup] platform: nordic_nrf: nRF54L15 does not have UICR
[nrf noup] platform: nordic_nrf: add support for nrf54l15dk
tf-m: Fix peripheral => base address for UART instance
tf-m: Add missing include for generic configurations
WIP: Adding the previous methodology of using base addresses
WIP: Fixing the last remaining issues for L15
WIP: TF-M: Fixing deprecated autoconf.h includes
[nrf noup] platform: nordic_nrf: Add support for nRF54L
[nrf noup] platform: nordic_nrf: Add support shared UART and using UART0 instance
[nrf noup] platform: nordic_nrf: Configure XL1/2 pin based on Kconfig
[nrf noup] platform: nordic_nrf: include autoconf.h from target_cfg.c
[nrf noup] platform: nrf_nordic: Add temp fix for for secure FPU on 9161
platform: nordic_nrf: restore TFM_DRIVER_STDIO's default value <-- This is a Zephyr noup

### CODE REVIEW

**Review by tomi-font — COMMENTED** [review]
- What commit is the squashed noup you are talking about?
- Is https://github.com/nrfconnect/sdk-trusted-firmware-m/pull/183/commits/5d5a7cb89138215e10127a52115db2eecdab3e68 more than just a cherry pick of the upstream PR ("sync repos")?
- Minor, but it would have been nice to include at least a small comment as to the why of the revert https://github.com/nrfconnect/sdk-trusted-firmware-m/pull/183/commits/dd5a627a447fbe3523492f3378e2da306901d91f.

**Comment by Vge0rge** [comment]
> * What commit is the squashed noup you are talking about?
I can provide a list of the noups which I squashed tomorrow that I will be in the office and will have access to the list again. But even when I provided that it is not that you can just squash them and get the exact same result because of conflict resolution etc.

> * Is [5d5a7cb](https://github.com/nrfconnect/sdk-trusted-firmware-m/commit/5d5a7cb89138215e10127a52115db2eecdab3e68) more than just a cherry pick of the upstream PR ("sync repos")?
What do you mean more than that just a cherry pick? 

> * Minor, but it would have been nice to include at least a small comment as to the why of the revert [dd5a627](https://github.com/nrfconnect/sdk-trusted-firmware-m/commit/dd5a627a447fbe3523492f3378e2da306901d91f).
I will add a comment for this in the PR description, yes.

**Comment by tomi-font** [comment]
> I can provide a list of the noups which I squashed tomorrow that I will be in the office and will have access to the list again. But even when I provided that it is not that you can just squash them and get the exact same result because of conflict resolution etc.

I was just wondering what commit is it that now has all the previous commits that are now squashes, cause I couldn't spot it with a quick look.

>  What do you mean more than that just a cherry pick?

Is that commit really just a cherry pick from upstream, or did you make additions to that commit which make it more than just the cherry pick of a single commit?

>  I will add a comment for this in the PR description, yes.

or in the commit :smiley:

**Comment by Vge0rge** [comment]
> > I can provide a list of the noups which I squashed tomorrow that I will be in the office and will have access to the list again. But even when I provided that it is not that you can just squash them and get the exact same result because of conflict resolution etc.
> 
> I was just wondering what commit is it that now has all the previous commits that are now squashes, cause I couldn't spot it with a quick look.

Oh right, I thought that you meant it the other way around. The fromlist commit is the result of squashing multiple noups and resolving conflicts, so this is what I am referring to here.


> 
> > What do you mean more than that just a cherry pick?
> 
> Is that commit really just a cherry pick from upstream, or did you make additions to that commit which make it more than just the cherry pick of a single commit?
So the commit that I have here is a clean cherry-pick from the upstream. That was the main goal of this task for me, to create a single commit that I can push upstream and I can also cherry pick in this repo cleanly.
 
> > I will add a comment for this in the PR description, yes.
> 
> or in the commit 😃

The reason that I didn't put it in the commit message is that the commit itself goes to upstream TF-M and thus noups from zephyr are not relevant. But I can add more generic statement in the commit message and a more specific one in the PR if that helps.

**Comment by tomi-font** [comment]
> The reason that I didn't put it in the commit message is that the commit itself goes to upstream TF-M and thus noups from zephyr are not relevant. But I can add more generic statement in the commit message and a more specific one in the PR if that help

Here I meant the revert commit (dd5a627a447fbe3523492f3378e2da306901d91f). It's not going anywhere right?

**Comment by Vge0rge** [comment]
The revert commit will not go anywhere, no. This content of the commit 3980d5 is included in the fromlist commit here.
So the plan is that in the next upmerge of tf-m the Zephyr nopup will be empty and it should get removed.

File: platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/config.cmake [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+#-------------------------------------------------------------------------------
+
+include(${PLATFORM_PATH}/common/nrf54l15/config.cmake)
+
+set(SECURE_UART1                        OFF         CACHE BOOL      "Enable secure UART1" FORCE)
+set(SECURE_UART30                       OFF         CACHE BOOL      "Enable secure UART1" FORCE)
+set(BL2                                 OFF         CACHE BOOL      "Enable secure UART1" FORCE)
```
Comment (tomi-font) [inline platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/config.cmake:12]: comment fails + no newline at end of file (and in `cpuarch.cmake`)

File: platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/partition/flash_layout.h [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __FLASH_LAYOUT_H__
+#define __FLASH_LAYOUT_H__
+
+/* Flash layout on NRF5340 Application MCU with BL2:
```
Comment (tomi-font) [inline platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/partition/flash_layout.h:21]: ahem 5340 (multiple times)

File: platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/services/src/tfm_platform_system.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+void tfm_platform_hal_system_reset(void)
+{
+	/* Reset the system */
+	NVIC_SystemReset();
+}
+
+
+/* These two functions need to be ported to the upstream project when support
+ * for nrf54l15 is added upstream. Remove these dummy implementations when these
+ * are ported. */
```
Comment (tomi-font) [inline platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/services/src/tfm_platform_system.c:22]: Something to be changed here maybe (at least the comment)?

File: platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/partition/flash_layout.h (hunk 2 above)
Comment (Vge0rge) [reply platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/partition/flash_layout.h:21]: Yeah, that makes sense. I write it in the commit message that most of the files are plain copies of the 5340 files just to allow to build. These will have to be basically ported to the nRF54L15 when we want to add support for it upstream.

File: platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/config.cmake (hunk 1 above)
Comment (Vge0rge) [reply platform/ext/target/nordic_nrf/nrf54l15dk_nrf54l15_cpuapp/config.cmake:12]: You are right that this is of course wrong,  and the same goes for your comment in the tfm_platform_system.c. I think that the upstream PR is about to get merged now. If it does I will just create another PR to fix these things upstream directly.

**Review by tomi-font — APPROVED** [review]
Acking as is assuming what needs fixing will be handled later.

**Comment by Vge0rge** [comment]
I have conflicts here because my intention is to force push to othe reposity. Instead of reverting hundrets of commits and applying them again I will try to get this PR force pushed in the repo.

This PR starts applying the nrf fromtree/noup commits starting from a clean (meaning from a commit from Zephyr's TF-M fork) history.

**Review by tomi-font — APPROVED** [review]

**Pull request merged**